juju-clean

```diff
@@ -17,55 +17,108 @@
+  do
+      sudo stop ${jujudaemon} || true
+  done
+  echo "Destroying juju template containers" >&2
+  for container in $(sudo lxc-ls | grep -e "^juju.*template$" )
+  do
+    sudo lxc-stop --name ${container} && sudo lxc-destroy --name ${container} || true
+  done
+
+  echo "Cleaning local providor files" >&2
+  sudo rm /etc/init/juju-* || true
+  sudo rm /etc/rsyslog.d/25-juju* || true
+  sudo rm -rf /var/lib/juju/containers/${USER}-* || true
+  sudo rm -rf /var/lib/juju/locks/* || true
+  echo "Cleaning lxc files" >&2
+  sudo rm -rf /var/cache/lxc/cloud-* || true
+  sudo rm /var/lib/lxc/juju-* || true
+  sudo rm /etc/lxc/auto/${USER}-* || true
+}
 
-echo "Destroying '$JUJU_ENV' environment" >&2
-juju destroy-environment -y $JUJU_ENV >/dev/null 2>&1 || true
-juju destroy-environment --force -y $JUJU_ENV >/dev/null 2>&1 || true
-
-echo "Removing any extra environment files in '${JUJU_HOME}'" >&2
-rm -f ~/${JUJU_HOME}/environments/$JUJU_ENV.jenv
-rm -rf ~/${JUJU_HOME}/$JUJU_ENV || true
-
-echo "Checking for errors with local provider" >&2
-
-echo "Stopping juju daemons" >&2
-for jujudaemon in $(sudo initctl list | grep juju | cut -d ' ' -f 1)
-do
-    sudo stop ${jujudaemon} || true
-done
-echo "Destroying juju template containers" >&2
-for container in $(sudo lxc-ls | grep -e "^juju.*template$" )
-do
-  sudo lxc-stop --name ${container} && sudo lxc-destroy --name ${container} || true
-done
-
-echo "Cleaning local providor files" >&2
-sudo rm /etc/init/juju-* || true
-sudo rm /etc/rsyslog.d/25-juju* || true
-sudo rm -rf /var/lib/juju/containers/${USER}-* || true
-sudo rm -rf /var/lib/juju/locks/* || true
-echo "Cleaning lxc files" >&2
-sudo rm -rf /var/cache/lxc/cloud-* || true
-sudo rm /var/lib/lxc/juju-* || true
-sudo rm /etc/lxc/auto/${USER}-* || true
-
-echo "Clean and ready for bootstrap"
+function clean_juju_2() {
+  local juju_controller=$1
+  # The juju 2.x configuration directory.
+  local juju_home=${JUJU_HOME:-~/.local/share/juju}
+  local description="Destroy and completely clean a juju 2.x controller."
+  local warning="This is an extremely destructive operation, born from the
+  frustrating situation where all other juju commands fail.\n
+  Try 'juju destroy-controller controller' first.\n
+  Then 'juju kill-controller controller' second.\n
+  If all else fails use clean, and may god have mercy on the controller because
+  this script will not."
+
+  if [ -z "${juju_controller}"]; then
+    echo $description
+    echo -e $warning
+    exit 1
+  fi
+
+  # Only kill lxd containers if you are trying to clean a lxd controller.
+  if [[ ${juju_controller} == *lxd* ]]; then
+    echo "Stopping and deleting the Juju LXD containers."
+    local containers=$(lxc list | grep juju | cut -d '|' -f 2)
+    for container in $containers ; do
+      lxc stop ${container}
+      lxc delete ${container}
+    done
+  else
+    echo "No local containers to clean for ${juju_controller}"
+  fi
+
+  echo "Cleaning local Juju 2.x configuration files."
```

Comment:
These next few steps are actually REALLY destructive. We don't want to just delete all controllers because LXD failed to kill.